common/attest_plugin.c

```diff
@@ -273,6 +306,119 @@ oe_result_t oe_verify_evidence(
+    if (!pub_key_buff)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    result = oe_cert_read_der(&cert, cert_in_der, cert_in_der_len);
+    OE_CHECK_MSG(result, "cert_in_der_len=%d", cert_in_der_len);
+
+    // validate the certificate signature
+    result = oe_cert_verify(&cert, NULL, NULL, 0);
+    OE_CHECK_MSG(
+        result,
+        "oe_cert_verify failed with error = %s\n",
+        oe_result_str(result));
+
+    //------------------------------------------------------------------------
+    // Validate the report's trustworthiness
+    //------------------------------------------------------------------------
+
+    // determine the size of the extension
+    if (oe_cert_find_extension(
+            &cert, (const char*)oid_oe_evidence, NULL, &report_size) !=
+        OE_BUFFER_TOO_SMALL)
+        OE_RAISE(OE_FAILURE);
+
+    report = (uint8_t*)oe_malloc(report_size);
+    if (!report)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    // find the extension
+    OE_CHECK(oe_cert_find_extension(
+        &cert, (const char*)oid_oe_evidence, report, &report_size));
+    OE_TRACE_VERBOSE("extract_x509_report_extension() succeeded");
+
+    // find the report version
+    header = (oe_report_header_t*)report;
+    if (header->version != OE_ATTESTATION_HEADER_VERSION)
+        OE_RAISE_MSG(OE_INVALID_PARAMETER, "Invalid report version", NULL);
+
+    result = oe_verify_evidence(
+        // rely on the format UUID in the header. For attestation
+        // certificate, the report should always include the header.
+        NULL,
+        report,
+        report_size,
+        NULL,
+        0,
+        NULL,
+        0,
+        &claims,
+        &claims_length);
+    OE_CHECK(result);
+    OE_TRACE_VERBOSE("quote validation succeeded");
+
+    // verify report data: hash(public key)
+    // extract public key from the cert
+    oe_memset_s(pub_key_buff, KEY_BUFF_SIZE, 0, KEY_BUFF_SIZE);
+    result =
+        oe_cert_write_public_key_pem(&cert, pub_key_buff, &pub_key_buff_size);
+    OE_CHECK(result);
+    OE_TRACE_VERBOSE(
+        "oe_cert_write_public_key_pem pub_key_buf_size=%d", pub_key_buff_size);
+
+    result = _verify_public_key_claim(
+        claims, claims_length, pub_key_buff, pub_key_buff_size);
+    OE_CHECK(result);
+    OE_TRACE_VERBOSE("user data: hash(public key) validation passed", NULL);
+
+    //---------------------------------------
+    // call client to further check claims
+    // --------------------------------------
+    if (claim_verify_callback)
+    {
+        result = claim_verify_callback(claims, claims_length, arg);
+        OE_CHECK(result);
+        OE_TRACE_VERBOSE("claim_verify_callback() succeeded");
+    }
+    else
+    {
+        OE_TRACE_WARNING(
+            "No claim_verify_callback provided in "
+            "oe_verify_attestation_certificate_with_evidence call",
+            NULL);
+    }
+
+done:
+    oe_free(pub_key_buff);
+    for (size_t i = 0; i < claims_length; i++)
```

Comment:
Consider using the `oe_free_claims` method instead

Reply:
Thanks. Done.